test/specs/validate-address.spec.js

```diff
@@ -154,38 +253,278 @@ describe("validateAddress()", () => {
+    let error;
+    const addressToValidate = {
+      country: "US",
+      street: ["4 Jersey St"],
+      cityLocality: "",
+      stateProvince: "",
+      postalCode: "",
+    };
+
+    try {
+      await shipengine.validateAddress(addressToValidate);
+    } catch (e) {
+      error = e;
+    }
+
+    expect(error).to.be.not.undefined;
+    expect(error.source).to.equal("shipengine");
+    expect(error.type).to.equal("validation");
+    expect(error.code).to.equal("field_value_required");
+    expect(error.message).to.equal(
+      "Invalid address. Either the postal code or the city/locality and state/province must be specified."
+    );
+    expect(error.requestId).to.be.undefined;
+  });
+
+  it("Throws an error if neither the postalCode nor the cityLocality is provided", async function () {
+    const shipengine = new ShipEngine({ apiKey, baseURL });
+    let error;
+    const addressToValidate = {
+      country: "US",
+      street: ["4 Jersey St"],
+      stateProvince: "MA",
+    };
+
+    try {
+      await shipengine.validateAddress(addressToValidate);
+    } catch (e) {
+      error = e;
+    }
+
+    expect(error).to.be.not.undefined;
+    expect(error.source).to.equal("shipengine");
+    expect(error.type).to.equal("validation");
+    expect(error.code).to.equal("field_value_required");
+    expect(error.message).to.equal(
+      "Invalid address. Either the postal code or the city/locality and state/province must be specified."
+    );
+    expect(error.requestId).to.be.undefined;
+  });
+
+  it("Throws an error if neither the postalCode nor the stateProvince is provided", async function () {
+    const shipengine = new ShipEngine({ apiKey, baseURL });
+    let error;
+    const addressToValidate = {
+      country: "US",
+      street: ["4 Jersey St"],
+      cityLocality: "Boston",
+    };
+
+    try {
+      await shipengine.validateAddress(addressToValidate);
+    } catch (e) {
+      error = e;
+    }
+
+    expect(error).to.be.not.undefined;
+    expect(error.source).to.equal("shipengine");
+    expect(error.type).to.equal("validation");
+    expect(error.code).to.equal("field_value_required");
+    expect(error.message).to.equal(
+      "Invalid address. Either the postal code or the city/locality and state/province must be specified."
+    );
+    expect(error.requestId).to.be.undefined;
+  });
+
+  it("Validates an address when a postalCode is provided but neither cityLocality nor stateProvince is provided", async function () {
+    const shipengine = new ShipEngine({ apiKey, baseURL });
+
+    const addressToValidate = {
+      country: "US",
+      street: ["4 Jersey St", "validate-residential-address"],
+      postalCode: "02215",
+    };
+
+    const response = await shipengine.validateAddress(addressToValidate);
+
+    const { normalizedAddress, isValid } = response;
+
+    expect(isValid).to.be.a("boolean").and.to.be.true;
+    expect(normalizedAddress.isResidential).to.be.a("boolean").and.to.be.true;
+
```

Comment:
This test should have assertions which verify that the city and state are populated in the normalized address even though they weren't specified in the input address.  It should also verify that the rest of the normalized address matches the input address.  You can use your `addressToValidate` + `expectedNormalizedAddress` pattern and the `assertNormalizedAddressMatchesOriginal()` function to do both of these.

Reply:
Gotcha!